unity-client/Assets/Plugins/CustomRequestAnimationFrame/CustomRequestAnimationFrame.jspre

```diff
@@ -0,0 +1,19 @@
+
+(function () {
+  // Taken from UnityInstallDirectory/PlaybackEngines/WebGLSupport/BuildTools//Emscripten/src/library_browser.js -> fakeRequestAnimationFrame(), and customized
+  window.requestAnimationFrame = function (rafCallback) {
+    // try to keep 30fps between calls to here
+    var now = Date.now();
+    var targetFPS = 30;
+    if (Browser.nextRAF === 0) {
+      Browser.nextRAF = now + (1000/targetFPS);
+    } else {
+      while (now + 2 >= Browser.nextRAF) { // fudge a little, to avoid timer jitter causing us to do lots of delay:0
+        Browser.nextRAF += (1000/targetFPS);
+      }
+    }
+
+    var delay = Math.max(Browser.nextRAF - now, 0);
+    setTimeout(rafCallback, delay);
```

Comment:
Warning: We shouldn't use setTimeout for frame execution.

See here: https://stackoverflow.com/questions/38709923/why-is-requestanimationframe-better-than-setinterval-or-settimeout.

If we want to cap requestAnimationFrame we should skip frames manually to produce the desired frame rate.

See here: https://stackoverflow.com/questions/19764018/controlling-fps-with-requestanimationframe (I suggest looking at all the answers. Not just the first one).

Reply:
If we are going to match 30 fps, maybe the best way to do this is to skip every other rAF so the intermediate calls do nothing.

I think that's is what's being referred in the Unity thread when talking about "decimating calls".

Reply:
I'll be fiddling a bit more with this today, I think i misread Unity's approach and they only use the "fake RAF" that has the setTimeout only the first time, or something like that (I interpreted this was used all the time and that's why I unified the background RAF that Pablo had implemented with this new override).

My objective is to have the browser native RAF at a frequency we can control (30FPS for this test, and should be able to set that framerate dynamically) in the .jspre file.

Let me try a few more things later today

Reply:
I think we should handle the framerate in unity directly. Unity has something called cooperative scheduling, it is the formal name of what is happening behind the "coroutines". It works by setting an execution quota every frame, and then continuing with important jobs (like actually rendering) once either the quota is complete or the "coroutines" are done.

By modifying the framerate in unity you may be modifying (_increasing_) the quota, therefore, affecting positively the CPU processing power per frame. 
In the contrary, by manually changing the external timer, you have the same effect as executing something else in the same thread (emulating another CPU intensive task), which has the effect of a non-preemptive scheduling schema, and therefore, it is not an optimization. It is instead, idle time in the CPU

Reply:
The comment may be hard to digest. We can schedule a call to better explain the concepts

Reply:
After fighting the whole day with this I finally achieved the decimation to get 30FPS using RequestAnimationFrame instead of setTimeout.

Check out the description of this PR as I updated it with the pertinent info, you can also check out the new build.

And of course, let's talk about that other approach tomorrow.

Reply:
@menduz I am quite sceptic of that cooperative scheduling you are mentioning. Coroutines just work by a simple nested `IEnumerator`  loop technique, and I don't think they throttle the processing of the `MoveNext()`. Already revised many many custom coroutine loop implementations and even implemented one myself, we even have a custom coroutine implementation that supports exceptions already in the project. 

In fact, all of our throttling code (and UnityGLTF and GLTFast throttling code) is laid under the assumption that the coroutines loop aren't being throttled automatically.

This said, I think we should compare the current rAF based implementation against the `setTimeout` one, because I just tested it and it seems to perform a bit worse. In the previous implementation I noted that we got almost permanent 30 fps, and in the current build I can see a bit more of 25-30 fps oscillation. Maybe an anomaly but i'd safe to check.

Reply:
To add a final note, please let's be wary of getting too technical there. We should go for what works regardless of the preached best practices.

Reply:
Ok, cool, let's continue this on slack. FWIW:
 - Preemptive scheduling: it is what you think I said. By the quota I mean that not all the Update() run all the frames. Please validate that assumption.
 - [Cooperative scheduling](https://en.wikipedia.org/wiki/Cooperative_multitasking) (what i mentioned): it is specifically the `yield return` behavior. It is not specific to unity, probably you will find better results searching for .Net